### ISSUE TITLE
Incompatible `self` types in method override

### ISSUE THREAD

**Issue by Timmmm** [microsoft/pyright#2805]
Consider this code (it's actually [a test case from MyPy](https://github.com/python/mypy/blob/master/test-data/unit/check-classes.test#L4170)):

```python
class B():
    def n(self, a: int) -> None: pass
class C(B):
    def m(self, a: int) -> None: pass
    n = m
```

Pyright reports an error for `n = m`:

    Expression of type "(self: Self@C, a: int) -> None" cannot be assigned to declared type "(self: Self@B, a: int) -> None"

I came across this when fixing [a bug in MyPy](https://github.com/python/mypy/issues/10375). My fix inadvertently cause that test case to fail by producing this error:

    Incompatible types in assignment (expression has type "Callable[[C, int], None]", variable has type "Callable[[B, int], None]")

(I think MyPy has some issues with `Callable` and class methods - sometimes it includes the type of `self` and sometimes it doesn't.) But anyway in this case it is a pretty similar error to Pyright!

So I starting thinking that I'd fixed another bug in MyPy, and based on your previous excellent help Pyright seems to be always right and MyPy is a complete mess.... but I can't actually figure out why the above code *should* fail type checking.

I've tried writing some code with the same type signatures that fails at runtime, but I can't come up with anything. Should it actually pass?

**Comment by erictraut** [comment]
Which version of pyright are you using? I think this is a bug I fixed several versions ago, and I'm no longer able to repro it with the latest.

**Comment by Timmmm** [comment]
Pylance v2021.12.2, which uses Pyright 1.1.195 according to its changelog.

**Comment by erictraut** [comment]
Yeah, I think this was fixed in pyright since then. Pyright is currently on version 1.1.203, so pylance is 8 versions out of date at this point. There should be a new release of pylance in the next few days. I'm going to close this for now, but if my assumption is incorrect, I'm happy to take another look.
